alacritty_terminal/src/term/text_run.rs

```diff
@@ -0,0 +1,309 @@
+    /// This is a destructive operation, the iterator will be in a new run state after it's
+    /// completion.
+    fn produce_cursor(&mut self, rc: RenderableCell) -> Option<TextRun> {
+        let (opt_start, _) = self.start_run(rc);
+        let start = opt_start?;
+        let cursor = self.cursor.take()?;
+        Some(TextRun::from_cursor_rc(start, cursor))
+    }
+
+    /// Create a run of chars from the current state of the `TextRunIter`.
+    /// This is a destructive operation, the iterator will be in a new run state after it's
+    /// completion.
+    fn produce_char_run(&mut self, rc: RenderableCell) -> Option<TextRun> {
+        let prev_buffer = self.drain_buffer();
+        let (start_opt, latest_opt) = self.start_run(rc);
+        let start = start_opt?;
+        let latest = latest_opt?;
+        Some(TextRun::from_iter_state(start, latest, prev_buffer))
+    }
+}
+
+impl<I> Iterator for TextRunIter<I>
+where
+    I: Iterator<Item = RenderableCell>,
+{
+    type Item = TextRun;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut output = None;
+        while let Some(rc) = self.iter.next() {
+            if self.latest.is_none() || self.run_start.is_none() {
+                // Initial state, this is should only be hit on the first next() call of
+                // iterator
+                self.run_start = Some(from_rc!(rc));
+            } else if self.cursor.is_some() {
+                // Last iteration of the loop found a cursor
+                // Return a run for the cursor and start a new run
+                output = self.produce_cursor(rc);
+                break;
+            } else if self.is_end_of_run(&rc) || rc.is_cursor() {
+                // If we find a run break or a cursor,
+                // return what we have so far and start a new run.
+                output = self.produce_char_run(rc);
+                break;
+            }
+            // Build up buffer and track the latest column we've seen
+            self.latest = Some((rc.column, rc.flags.contains(Flags::WIDE_CHAR)));
+            self.buffer_content(rc.inner);
+        }
+        // If we generated output we're done.
+        // Otherwise check for any remaining buffered content and return it as a text run.
+        // This is a destructive operation, it will return None after it excutes once.
+        output.or_else(|| {
+            if !self.buffer_text.is_empty() || !self.buffer_zero_width.is_empty() {
+                let start = self.run_start.take()?;
+                let latest = self.latest.take()?;
+                // Save leftover buffer and empty it
+                Some(TextRun::from_iter_state(start, latest, self.drain_buffer()))
+            } else if let Some(cursor) = self.cursor {
+                let start = self.run_start.take()?;
+                Some(TextRun::from_cursor_rc(start, cursor))
+            } else {
+                None
+            }
+        })
+    }
```

Comment:
Okay, so the `TextRunIterator` functions are a bit messy I think. I don't have a clear instruction here on how to improve things since that would require a lot of experimenting on my side too, but to try to get you some information about what needs to be changed I'll list some issues I see with the current approach:

```
buffer_content -> Why do we set cursor here to notify next col to start a new run?
produce_cursor -> Do we really need to create a new run with relatively generic logic? Wouldn't it be easier to just handle this separately as soon as a cursor is encountered?
produce_char_run -> Why `prev_buffer` isn't it the current buffer? We first collect all text, then start a new run, then construct a new run from the last run? That's a bit confusing.
start_run -> Does this return the info of the last run? A bit unclear and not documented in name or comments, there's probably a better solution.
TextRunIterator::new
    -> It seems like it would be much easier to early return rather than buffering `output`
    -> output.or_else is used to return the last text run? I think this is a bit confusing and caused by the start_run being unclear too.
    -> The whole cursor handling in output.or_else could probably be handled by explicitly handling cursors separately
```

Of course these are very general statements without clear instructions, so please let me know when you have any more questions. But I think it should be pretty clear here that control flow is not straight forward and I think we can do better.

Reply:
> At a high level I believe a lot of the issue here spawn from the fact that we have to destructively grab enough RenderableCells from our internal iterator.

Maybe we could use `peek`?

I can take a deeper look at it and rewrite things myself, though that would take time I currently do not have (at least for this PR). But I'm pretty confident we can improve things a lot here.